src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LoadedProject.cs

```diff
@@ -95,6 +101,22 @@ private void ProjectFileChangeContext_FileChanged(object? sender, string filePat
         NeedsReload?.Invoke(this, EventArgs.Empty);
     }
 
+    private void AssetsFileChangeContext_FileChanged(object? sender, string filePath)
+    {
+        // We only want to trigger design time build if the assets file content actually changed from the last time this handler was called.
+        // Sometimes we can get a change event where no content changed (e.g. for a failed restore).
+        // In such cases, proceeding with design-time build can put us in a restore loop (since the design-time build notices that assets are missing).
+        using var assetsFileStream = File.OpenRead(filePath);
```

Comment:
Consider making this async.

Reply:
Which operation should be made async? I did not see a `File.OpenAsync` or similar method. I did find discussion https://github.com/dotnet/runtime/discussions/63153 though.

Reply:
Oh I didn't realize that wasn't a thing. Never mind then.